Some changes for better code quality

Levev opened the pull request:
I've come across this project on reddit while I was burnt out, so I looked into the source code, and I noticed some things that could be easily changed for extra readability, so I made this pull request. This pr doesn't aim to dig into every nook and cranny of the source/change everything that can be changed, rather I just wanted to show some things that could've been done better. I hope someone will find this useful in the end.

Levev committed:
Dynamic allocation cleanup of dynarray

Some things just shouldn't be put on the heap. This commit aims to fix this quote on quote issue. I discovered this little project on reddit, and decided to give it a look. The first file I opened was dynarray.c, and I got a little spooked for a second. I got rid of some unnecessary heap allocs, while trying to be as close to the original style of the codebase as possible

Levev committed:
Use references only in the right situation

I've come across a lot of function signatures in this codebase that ask for a reference to something, without the need of mutability. I tried fixing these, but the codebase is a little hard to follow, especially when it comes to `board.c`. In the end I hope this commit communicates what my intention was.

Levev committed:
Fix arguments in `io`

Chewt commented:
Hi,

First off, thank you for taking the time to look through the source code of my project, it brings me joy to see that it is being seen by others. 

I have a few issues with your pr however, and would like to discuss them here.
First, readability: After reviewing your changes, I am not entirely sure how you made the code more readable. In some cases you actually took away readability by removing the portions of code that identified which piece of the board was being checked against (see line 73 of board.c). Other than that your changes made some struct pointers that were passed to functions into the structs themselves, which then changed all of the arrow operators into dot operators. I personally don't think dot operators are more readable than arrow operators, or vice versa for that matter. There is a bigger issue with changing these pointers into values however...

The decision to pass structs via pointer was deliberate. Whenever you pass a variable to a function, a copy of that variable is made on the stack for use in that function. For structs, this can lead to large memory usage, and reduced code performance due to time spent copying data onto the stack. There is also the issue of deep vs shallow copying of structs. For example lets say my struct contains a couple of ints, and a pointer to an array of ints. Passing this struct to a function will create a new struct with a copy of all of those ints, and a copy of the memory address. This means that I can modify that struct however I want and it will not affect the original struct, except in the case of that pointer. That pointer still points to the same array of ints as the original, which means that If I were to modify that array, then use the original someplace else it will have entirely different values. This side effect is difficult to keep track of and can lead to bugs later on. I notice you removed the deep copy I had in place on line 207 of board.c, which leads me to believe you were not aware of the difference between deep and shallow copies, which is why I went through the trouble of describing them in detail. 

I apologize if I explained some concepts that you are already knowledgeable about, I just wanted to make clear my issues with this pull request in its current state.

I know how difficult burnout can be, so I hope that this effort you put into my project has lifted you out of burnout and back into productivity. Feel free to contribute to this project on other ways if you see anything you think you can improve, and I would be happy to look it over once more.

Levev committed:
Patch things up

I removed some debug characters that I accidentally left in the code. I also noticed some explicit null characters in string literals that didn't serve a purpose, so I removed those as well.

Levev commented:
Hey, thanks for taking your time to look into the changes I made, I appreciate it. However, I wanted to give an answer/explanation to some of your concerns with this pull request.
>  Other than that your changes made some struct pointers that were passed to functions into the structs themselves, which then changed all of the arrow operators into dot operators. I personally don't think dot operators are more readable than arrow operators, or vice versa for that matter.

I like to make my function signatures as expressive as possible. For example, if you look at a function that takes a pointer to some data as an argument, you can assume that given function will make changes to the data it asks a pointer for. You could say that the function asks for mutability. So I went ahead and looked at some of your functions that take a pointer as a parameter to see if they actually need the mutability that a pointer provides or not. 

> The decision to pass structs via pointer was deliberate. Whenever you pass a variable to a function, a copy of that variable is made on the stack for use in that function. For structs, this can lead to large memory usage, and reduced code performance due to time spent copying data onto the stack.

I can't see why copying data onto the stack would be slower than to memcpy data from an arbitrary buffer into a stack allocated variable. The only issue with passing bigger structs by value would be the stack size itself, but that isn't that big of a concern, given the size of the struct in question.

>  There is also the issue of deep vs shallow copying of structs.

The difference between deep and shallow copying is insignificant if the struct itself doesn't contain/own pointers. Let's look at the definition of the [`Board`](https://github.com/Levev/Chessterm/blob/9c8f55019fe89aa384b3a919bb1081c2dff3dd46/include/board.h#L50) struct as an example, as most of my changes revolve around it. It doesn't contain any pointers. It does, however, contain static arrays, which are part of the struct itself, meaning that the contents of these arrays get copied upon assignment.

I didn't make this pull request with the goal of it being merged. I just wanted to give some ideas the project could benefit from later on, and I hope I was able to give some insight into my philosophy when it comes to coding. I wish you the best with this/future projects!

Chewt commented:
> I like to make my function signatures as expressive as possible. For example, if you look at a function that takes a pointer to some data as an argument, you can assume that given function will make changes to the data it asks a pointer for. You could say that the function asks for mutability. So I went ahead and looked at some of your functions that take a pointer as a parameter to see if they actually need the mutability that a pointer provides or not.

Personally I would prefer to mark the pointer as const rather than pass the struct by value in order to express immutability. This is because I would rather only to a local copy of a pointer to a variable within the function's scope rather than a copy of the whole struct. I also believe the visible `const` keyword does a better job to express my intentions with the data. I obviously didn't do that in this project where I could have, so I think I'll go back and add it to the functions that don't actually change the structs values. 

> I can't see why copying data onto the stack would be slower than to memcpy data from an arbitrary buffer into a stack allocated variable. The only issue with passing bigger structs by value would be the stack size itself, but that isn't that big of a concern, given the size of the struct in question.

There are only a handful of places where I use memcpy in a function. I could change those functions to pass by value and save a couple of lines, but as you say it would probably not be any faster or slower than keeping it how it is. I also want to have consistency in my function signatures.

> The difference between deep and shallow copying is insignificant if the struct itself doesn't contain/own pointers. Let's look at the definition of the [`Board`](https://github.com/Levev/Chessterm/blob/9c8f55019fe89aa384b3a919bb1081c2dff3dd46/include/board.h#L50) struct as an example, as most of my changes revolve around it. It doesn't contain any pointers. It does, however, contain static arrays, which are part of the struct itself, meaning that the contents of these arrays get copied upon assignment.

I will admit that It had been a while and I definitely was remembering my other chess project which has structs with pointers to memory in them. You make a valid point on the deep vs shallow copy, my apologies.

I'm going to close this PR, since you don't intend to have it pulled. Thank you for discussing this with me, it's fun to talk about a personal project with someone else! I'll wish you the same on future projects!
